src/TTextEdit.cpp

```diff
@@ -983,37 +989,77 @@ void TTextEdit::mouseMoveEvent(QMouseEvent* event)
+        // beginning lines deleted!
+
+        // Count of "normal" width equivalent characters - we will multiply that
+        // by the average character width to determine whether the mouse is over
+        // a particular grapheme:
+        int column = 0;
+        // This is the calculated (estimated) cumulative width of all of the
+        // graphemes considered in the line so far:
         int currentX = 0;
-        for (const auto& character : mpBuffer->lineBuffer.at(lineNumber)) {
-            const uint unicode = getGraphemeBaseCharacter(character);
+        QString lineText = mpBuffer->lineBuffer.at(lineNumber);
+        QTextBoundaryFinder boundaryFinder(QTextBoundaryFinder::Grapheme, lineText);
+#if defined (QT_DEBUG)
+        if (debug_mousePositionInLine) {
+            qDebug().noquote().nospace() << "TTextEdit::convertMouseXToBufferX(" << mouseX << ", " << lineNumber << ") INFO - finding mouse click position in line:";
+        }
+#endif
+        for (int indexOfChar = 0, total = lineText.size(); indexOfChar < total; /*incrementing is done inside loop*/) {
+            int nextBoundary = boundaryFinder.toNextBoundary();
+            // Width in "normal" width equivalent of this grapheme:
+            int charWidth = 0;
+            const QString grapheme = lineText.mid(indexOfChar, nextBoundary - indexOfChar);
+#if defined (QT_DEBUG)
+            if (debug_mousePositionInLine) {
+                if (indexOfChar > 0) {
+                    // Include graphemes BEFORE the current last one considered:
+                    qDebug().noquote().nospace() << lineText.left(indexOfChar) << "»" << grapheme << "«";
+                } else {
+                    // First grapheme:
+                    qDebug().noquote().nospace() << "»" << grapheme << "«";
+                }
+            }
+#endif
+            const uint unicode = getGraphemeBaseCharacter(grapheme);
             if (unicode == '\t') {
-                characterWidth = 8;
+                charWidth = mTabStopwidth - (column % mTabStopwidth);
             } else {
-                characterWidth = getGraphemeWidth(unicode);
+                charWidth = getGraphemeWidth(unicode);
             }
-            currentX += characterWidth * mFontWidth;
-            characterIndex++;
+            column +=charWidth;
+            currentX = column * mFontWidth;
             if (currentX >= mouseX) {
                 if (mShowTimeStamps) {
-                    characterIndex -= 13;
+                    // The timestamp is (currently) 13 "normal width" characters
+                    // but that might not always be the case in some future I18n
+                    // situations:
+                    return std::max(0, indexOfChar - 13);
                 }
-                characterIndex = std::max(characterIndex, 0);
-                return characterIndex;
+                return std::max(0, indexOfChar);
             }
+            // Increment for loop:
```

Comment:
This is not confusing nor difficult to understand - please, please let's not have such trivial comments such as "increment for loop".

Reply:
It was also a reminder for me to **ensure** that I put it in after noting that I was NOT doing it in the (missing third part of the) `for (..,; ...; ...)` ! AFAICR It needs to be done in that way in order to sequences the updating of the start and end `QChar` indexes so that they correctly select the right ones to form a *single* grapheme.

Reply:
Ok, but it's in now, we don't need it anymore.

Reply:
Okay, will be gone in next commit.